File: apps/capture/expo-graft-env/android/src/main/java/expo/modules/graftenv/ExpoGraftEnvModule.kt
package expo.modules.graftenv

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoGraftEnvModule : Module() {
  companion object {
    init {
      try {
        System.loadLibrary("expograftenv")
      } catch (e: Exception) {
        android.util.Log.e("ExpoGraftEnv", "Failed to load native library", e)
      }
    }
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoGraftEnv")

    // Defines event names that the module can send to JavaScript.
    Events("onConfigChange")

    /**
     * Set environment variables for graft extension
     * On Android, we use JNI to call native setenv() function
     * This is necessary because graft extension reads configuration from environment
     */
    AsyncFunction("setEnvironmentVariables") { config: Map<String, String?> ->
      try {
        var successCount = 0
        var errorMessage: String? = null
        
        config.forEach { (key, value) ->
          if (value != null) {
            val result = nativeSetEnv(key, value)
            if (result == 0) {
              successCount++
            } else {
              errorMessage = "Failed to set $key (error code: $result)"
            }
          }
        }
        
        if (errorMessage != null) {
          sendEvent("onConfigChange", mapOf(
            "success" to false,
            "message" to errorMessage
          ))
          throw Exception(errorMessage)
        } else {
          sendEvent("onConfigChange", mapOf(
            "success" to true,
            "message" to "Set $successCount environment variables successfully"
          ))
        }
      } catch (e: Exception) {
        sendEvent("onConfigChange", mapOf(
          "success" to false,
          "message" to "Failed to set environment variables: ${e.message}"
        ))
        throw e
      }
    }

    /**
     * Get an environment variable value
     */
    AsyncFunction("getEnvironmentVariable") { key: String ->
      nativeGetEnv(key)
    }

    /**
     * Clear all graft-related environment variables
     */
    AsyncFunction("clearEnvironmentVariables") {
      try {
        val graftKeys = listOf(
          "GRAFT_CONFIG",
          "AWS_ACCESS_KEY_ID",
          "AWS_SECRET_ACCESS_KEY",
          "AWS_REGION",
          "AWS_ENDPOINT"
        )
        
        var successCount = 0
        graftKeys.forEach { key ->
          val result = nativeUnsetEnv(key)
          if (result == 0) {
            successCount++
          }
        }
        
        sendEvent("onConfigChange", mapOf(
          "success" to true,
          "message" to "Cleared $successCount environment variables"
        ))
      } catch (e: Exception) {
        sendEvent("onConfigChange", mapOf(
          "success" to false,
          "message" to "Failed to clear environment variables: ${e.message}"
        ))
        throw e
      }
    }
  }

  /**
   * Native method to set environment variable using setenv()
   * Returns 0 on success, -1 on failure
   */
  private external fun nativeSetEnv(key: String, value: String): Int

  /**
   * Native method to get environment variable using getenv()
   * Returns null if not found
   */
  private external fun nativeGetEnv(key: String): String?

  /**
   * Native method to unset environment variable using unsetenv()
   * Returns 0 on success, -1 on failure
   */
  private external fun nativeUnsetEnv(key: String): Int
}

File: apps/capture/expo-graft-env/ios/ExpoGraftEnvModule.swift
import ExpoModulesCore
import Foundation

public class ExpoGraftEnvModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoGraftEnv")

    // Defines event names that the module can send to JavaScript.
    Events("onConfigChange")

    /**
     * Set environment variables for graft extension
     * On iOS, we use setenv() to set environment variables in the process
     */
    AsyncFunction("setEnvironmentVariables") { (config: [String: String?]) in
      do {
        for (key, value) in config {
          if let unwrappedValue = value {
            self.setEnvironmentVariable(key: key, value: unwrappedValue)
          }
        }
        
        // Send success event
        self.sendEvent("onConfigChange", [
          "success": true,
          "message": "Environment variables set successfully"
        ])
      } catch {
        self.sendEvent("onConfigChange", [
          "success": false,
          "message": "Failed to set environment variables: \(error.localizedDescription)"
        ])
        throw error
      }
    }

    /**
     * Get an environment variable value
     */
    AsyncFunction("getEnvironmentVariable") { (key: String) -> String? in
      guard let value = getenv(key) else {
        return nil
      }
      return String(cString: value)
    }

    /**
     * Clear all graft-related environment variables
     */
    AsyncFunction("clearEnvironmentVariables") {
      do {
        let graftKeys = [
          "GRAFT_CONFIG",
          "AWS_ACCESS_KEY_ID",
          "AWS_SECRET_ACCESS_KEY",
          "AWS_REGION",
          "AWS_ENDPOINT"
        ]
        
        for key in graftKeys {
          self.removeEnvironmentVariable(key: key)
        }
        
        self.sendEvent("onConfigChange", [
          "success": true,
          "message": "Environment variables cleared"
        ])
      } catch {
        self.sendEvent("onConfigChange", [
          "success": false,
          "message": "Failed to clear environment variables: \(error.localizedDescription)"
        ])
        throw error
      }
    }
  }

  /**
   * Set an environment variable using setenv()
   * This makes the variable available to child processes and C libraries
   */
  private func setEnvironmentVariable(key: String, value: String) {
    setenv(key, value, 1)
  }

  /**
   * Remove an environment variable using unsetenv()
   */
  private func removeEnvironmentVariable(key: String) {
    unsetenv(key)
  }
}
